components/BottomNavigation/src/MDCBottomNavigationView.m

```diff
@@ -0,0 +1,258 @@
+                    context:nil];
+    [tabBarItem addObserver:self
+                 forKeyPath:kMDCBottomNavigationViewBadgeValueString
+                    options:NSKeyValueObservingOptionNew
+                    context:nil];
+    [tabBarItem addObserver:self
+                 forKeyPath:kMDCBottomNavigationViewImageString
+                    options:NSKeyValueObservingOptionNew
+                    context:nil];
+    [tabBarItem addObserver:self
+                 forKeyPath:kMDCBottomNavigationViewTitleString
+                    options:NSKeyValueObservingOptionNew
+                    context:nil];
+  }
+}
+
+- (void)removeObserversFromTabBarItems {
+  for (UITabBarItem *tabBarItem in self.navBarItems) {
+    @try {
+      [tabBarItem removeObserver:self forKeyPath:kMDCBottomNavigationViewBadgeColorString];
+      [tabBarItem removeObserver:self forKeyPath:kMDCBottomNavigationViewBadgeValueString];
+      [tabBarItem removeObserver:self forKeyPath:kMDCBottomNavigationViewImageString];
+      [tabBarItem removeObserver:self forKeyPath:kMDCBottomNavigationViewTitleString];
+    }
+    @catch (NSException *exception) {
+      if (exception) {
+        // No need to do anything if there are no observers.
+      }
+    }
+  }
+}
+
+- (void)removeBottomNavigationCells {
+  for (MDCBottomNavigationCell *cell in self.navBarCells) {
+    [cell removeFromSuperview];
+  }
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
+                       context:(void *)context {
+  if (!context) {
+    NSInteger i = 0;
+    NSInteger selectedItemNum = 0;
+    for (UITabBarItem *tabBarItem in self.navBarItems) {
+      if (object == tabBarItem) {
+        selectedItemNum = i;
+        break;
+      }
+      i++;
+    }
+    MDCBottomNavigationCell *cell = _navBarCells[selectedItemNum];
+    if ([keyPath isEqualToString:kMDCBottomNavigationViewBadgeColorString]) {
+      [cell setBadgeColor:change[kMDCBottomNavigationViewNewString]];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationViewBadgeValueString]) {
+      [cell setBadgeValue:change[kMDCBottomNavigationViewNewString]];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationViewImageString]) {
+      [cell setImage:change[kMDCBottomNavigationViewNewString]];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationViewTitleString]) {
+      [cell setTitle:change[kMDCBottomNavigationViewNewString]];
+    }
+  }
+}
+
+- (void)didTapButton:(UIButton *)button {
+  for (MDCBottomNavigationCell *cell in self.navBarCells) {
+    if (cell.button != button) {
+      [cell setSelected:NO animated:YES];
+    }
+  }
+  MDCBottomNavigationCell *cell = (MDCBottomNavigationCell *)button.superview;
+  [cell setSelected:YES animated:YES];
+}
+
+- (void)selectItem:(UITabBarItem *)item {
+  NSInteger i = 0;
+  for (UITabBarItem *tabBarItem in self.navBarItems) {
```

Comment:
If we are using the counter as an index into another array, consider using a for loop with an explicit index.
for (NSInteger itemIndex = 0; itemIndex < self.navBarItems.count; ++itemIndex) {
  if (item == self.navBarItems[itemIndex]) {
    self.navBarCell[itemIndex].selected = YES;
  } else {
    self.navBarCell[itemIndex].selected = NO;
  }
}